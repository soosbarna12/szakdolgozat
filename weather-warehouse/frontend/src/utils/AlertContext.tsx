import React, { createContext, useState, useContext, ReactNode } from "react";
import { Snackbar, Alert } from "@mui/material";

type AlertContextType = {
    showAlert: (message: string, severity?: "error" | "info" | "success" | "warning") => void;
};

export const AlertContext = createContext<AlertContextType>({
    showAlert: () => { },
});

export function AlertProvider({ children }: { children: ReactNode }) {
    const [open, setOpen] = useState(false);
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState<"error" | "info" | "success" | "warning">("info");

    function showAlert(newMessage: string, newSeverity: "error" | "info" | "success" | "warning" = "info") {
        setMessage(newMessage);
        setSeverity(newSeverity);
        setOpen(true);
    }

    return (
        <AlertContext.Provider value={{ showAlert }}>
            {children}
            <Snackbar open={open} autoHideDuration={4000} onClose={() => setOpen(false)}>
                <Alert severity={severity} onClose={() => setOpen(false)}>
                    {message}
                </Alert>
            </Snackbar>
        </AlertContext.Provider>
    );
}

export const useAlert = () => useContext(AlertContext);