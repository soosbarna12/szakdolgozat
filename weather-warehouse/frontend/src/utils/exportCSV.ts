import { download, generateCsv, mkConfig } from "export-to-csv";
import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function exportCSV(data: HistoricalDataTable[]) {
  if (!data || data.length === 0) {
    return;
  }

  const csvConfig = mkConfig({ useKeysAsHeaders: true, filename: "historical_data" });

  const csv = generateCsv(csvConfig)(data as any[]);
  download(csvConfig)(csv);
}