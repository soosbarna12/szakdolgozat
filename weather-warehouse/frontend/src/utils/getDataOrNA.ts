export function getDataOrNA(data: any) {
  if (data) {
    return data;
  }
  return "N/A";
}