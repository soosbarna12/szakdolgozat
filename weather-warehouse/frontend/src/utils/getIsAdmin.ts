import { jwtDecode } from "jwt-decode";
import { JwtPayload } from "../types/jwtPayload";


export function getIsAdmin(): boolean {
    const token = localStorage.getItem('token');
    if (!token) return false;
    try {
      const decoded = jwtDecode<JwtPayload>(token);
      return decoded.role === 'admin';
    } catch {
      return false;
    }
  }