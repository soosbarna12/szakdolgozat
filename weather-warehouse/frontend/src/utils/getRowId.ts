import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function getRowId(row: HistoricalDataTable): string {
  return row.date + row.cityName + row.countryCode;
}