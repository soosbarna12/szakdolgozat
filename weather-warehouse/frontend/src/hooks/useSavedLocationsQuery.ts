import { useQuery } from "@tanstack/react-query";
import { useEffect } from "react";
import { SavedUserLocation } from "../types/historicalDataTable.type";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useSavedLocationQuery(open: boolean) {
  const { showAlert } = useAlert();
  
  const { data: savedLocations, error, isLoading, refetch } = useQuery({
    queryKey: ["fetchSavedLocations", open],
    queryFn: async () => {
      const response = await axios.get(`/api/user/fetchSavedLocations`);
      return response.data as SavedUserLocation[];
    },
    enabled: open,
    retry: false,
  });
  
  useEffect(() => {
    if (error) {
      showAlert("Error fetching saved locations", "error");
    }
  }, [error]);

  return { savedLocations, error, isLoading, refetch };
}
