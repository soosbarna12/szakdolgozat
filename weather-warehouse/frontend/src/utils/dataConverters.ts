import dayjs from "dayjs";
import { HistoricalDataTable, ServerHistoricalData } from "../types/historicalDataTable.type";
import { HistoricalLocationData } from "../contexts/HistoricalContext/HistoricalContext.type";
import { getDataOrNA } from "./getDataOrNA";

export function convertKelvinToCelsius(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(kelvin - 273.15);
}

export function convertKelvinToFahrenheit(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(((kelvin - 273.15) * 9) / 5 + 32);
}

export function convertCelsiusToFahrenheit(celsius: number | null): number | string {
	if (celsius === null) {
		return 'N/A';
	}
	return Math.round((celsius * 9) / 5 + 32);
}

export function convertFahrenheitToCelsius(fahrenheit: number | null): number | string {
	if (fahrenheit === null) {
			return 'N/A';
	}
	return Math.round((fahrenheit - 32) * 5 / 9);
}

export function convertTitleCase(str: string): string {
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

export function convertServerHistoricalData(historicalData: ServerHistoricalData[], location: HistoricalLocationData): HistoricalDataTable[] {
	return historicalData.map((data) => ({
		lat: location.lat,
		lon: location.lon,
		locationKey: data.LocationKey,
		dateKey: data.DateKey,
		date: dayjs(data.FullDate).format("YYYY-MM-DD"),
		cityName: data.CityName,
		countryCode: data.CountryCode,
		temp: getDataOrNA(data.Temperature),
		maxTemp: getDataOrNA(data.MaxTemperature),
		minTemp: getDataOrNA(data.MinTemperature),
		humidity: getDataOrNA(data.Humidity),
		windSpeed: getDataOrNA(data.WindSpeed),
		precipitation: getDataOrNA(data.Precipitation),
		pressure: getDataOrNA(data.Pressure),
		cloudCover: getDataOrNA(data.CloudCover),
	}));
}
