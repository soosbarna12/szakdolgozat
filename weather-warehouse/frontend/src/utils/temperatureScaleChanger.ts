import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../consts/temperatureScale.const";
import { TemperatureScale } from "../types/temperatureScale.type";
import { convertCelsiusToFahrenheit, convertFahrenheitToCelsius, convertKelvinToCelsius, convertKelvinToFahrenheit } from "./dataConverters";


export function temperatureScaleChanger(currentTempScale: TemperatureScale, goalTempScale: TemperatureScale, tempValue: number | string) : string | number {

  if (typeof tempValue === "string") {
    return tempValue;
  }

  // c to c / f to f
  if (goalTempScale === currentTempScale) {
    return tempValue;
  }
  if (goalTempScale === TemperatureScale.Celsius && currentTempScale === TemperatureScale.Fahrenheit) {
    return convertFahrenheitToCelsius(tempValue);
  }
  if (goalTempScale === TemperatureScale.Fahrenheit && currentTempScale === TemperatureScale.Celsius) {
    return convertCelsiusToFahrenheit(tempValue);
  }
  if (goalTempScale === TemperatureScale.Celsius && currentTempScale === TemperatureScale.Kelvin) {
    return convertKelvinToCelsius(tempValue);
  }
  if (goalTempScale === TemperatureScale.Fahrenheit && currentTempScale === TemperatureScale.Kelvin) {
    return convertKelvinToFahrenheit(tempValue);
  }

  return "N/A";
}
