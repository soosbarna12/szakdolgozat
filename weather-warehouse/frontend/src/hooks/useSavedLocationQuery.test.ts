import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import { useAlert } from '../utils/AlertContext';
import axios from '../utils/axiosConfig';
import { renderHookWithQueryClient } from '../utils/test/renderHookWithQueryClient';
import { useSavedLocationQuery } from './useSavedLocationsQuery';

jest.mock('../utils/AlertContext', () => ({
  useAlert: jest.fn(),
}));

jest.mock('../utils/axiosConfig');

describe('useSavedLocationQuery', () => {
  const mockShowAlert = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    (axios.get as jest.Mock).mockReset();

    (useAlert as jest.Mock).mockImplementation(() => ({
      showAlert: mockShowAlert,
    }));
  });

  it('matches useSavedLocationQuery hook snapshot', async () => {
    (axios.get as jest.Mock).mockReturnValue(Promise.resolve({ data: [] }));

    const { result } = renderHookWithQueryClient(() => useSavedLocationQuery(true));

    await waitFor(() => {
      expect(result.current.savedLocations).toEqual([]);
    });

    expect(result.current).toMatchSnapshot();
  });

  it('fetches saved locations successfully', async () => {
    const mockSavedLocations = [
      { id: 1, name: 'New York', lat: 40.7128, lon: -74.006 },
      { id: 2, name: 'Los Angeles', lat: 34.0522, lon: -118.2437 },
    ];
    (axios.get as jest.Mock).mockReturnValue(Promise.resolve({ data: mockSavedLocations }));

    const { result } = renderHookWithQueryClient(() => useSavedLocationQuery(true));

    await waitFor(() => {
      expect(result.current.savedLocations).toEqual(mockSavedLocations);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

it('shows error alert when fetching saved locations fails', async () => {
  // Mock axios to reject with an error
  (axios.get as jest.Mock).mockReturnValue(Promise.reject(new Error('Error fetching saved locations')));

  // Render the hook
  const { result } = renderHookWithQueryClient(() => useSavedLocationQuery(true));

  // Debugging log to verify axios mock calls
  console.log('axios.get mock calls:', (axios.get as jest.Mock).mock.calls);

  // Wait for the error state to be set
  await waitFor(() => {
    expect(result.current.error).not.toBeNull();
  });

  // Assert that the alert was shown
  await waitFor(() => {
    expect(mockShowAlert).toHaveBeenCalledWith('Error fetching saved locations', 'error');
  });
});

  it('does not fetch data when "open" is false', async () => {
    const { result } = renderHookWithQueryClient(() => useSavedLocationQuery(false));

    expect(result.current.savedLocations).toBeUndefined();
    expect(result.current.isLoading).toBe(false);
    expect(result.current.error).toBeNull();
  });
});