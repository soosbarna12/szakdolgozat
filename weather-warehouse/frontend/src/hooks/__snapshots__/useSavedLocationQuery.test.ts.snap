// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSavedLocationQuery matches useSavedLocationQuery hook snapshot 1`] = `
{
  "error": null,
  "isLoading": false,
  "refetch": [Function],
  "savedLocations": [],
}
`;
